File: services/http-gateway/app/services/rate_limiter.rb
# frozen_string_literal: true

class RateLimiter
  def initialize(redis: RedisClient.instance)
    @redis = redis
  end

  def allow?(key, limit:, window_ms:)
    redis_key = "rl:#{key}"
    count = @redis.incr(redis_key)
    @redis.pexpire(redis_key, window_ms) if count == 1
    count <= limit
  end
end

File: services/core/src/main/kotlin/com/sellbot/core/http/redis/RateLimiter.kt
package com.sellbot.core.http.redis

import org.springframework.data.redis.core.StringRedisTemplate
import org.springframework.stereotype.Component
import java.time.Duration

@Component
class RateLimiter(
    private val redis: StringRedisTemplate,
) {
    fun allow(key: String, limit: Int, windowMs: Long): Boolean {
        val redisKey = "rl:$key"
        val count = redis.opsForValue().increment(redisKey) ?: return false
        if (count == 1L) {
            redis.expire(redisKey, Duration.ofMillis(windowMs))
        }
        return count <= limit
    }
}
